[BEAM-13] Add JmsIO

jbonofre opened the pull request:
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [X] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [X] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [X] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [X] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---
JmsIO (unbounded source and sink).

jbonofre commented:
This is a first draft for discussion and improvements. I will improve with new commits.

R: @davorbonaci 
R: @dhalperi

davorbonaci commented on sdks/java/io/jms/pom.xml:
```diff
@@ -0,0 +1,98 @@
+    <dependency>
+      <groupId>org.apache.geronimo.specs</groupId>
+      <artifactId>geronimo-jms_1.1_spec</artifactId>
+      <version>1.1.1</version>
```
Move version to the top-level?

jbonofre replied:
As it's JMS specific, I think it makes sense to keep it local to the IO. WDYT ?

davorbonaci commented on sdks/java/io/jms/pom.xml:
```diff
@@ -0,0 +1,98 @@
+    <dependency>
+      <groupId>org.apache.activemq</groupId>
+      <artifactId>activemq-broker</artifactId>
+      <version>5.11.2</version>
```
Move version to the top-level?

jbonofre replied:
As it's JMS specific, I think it makes sense to keep it local to the IO. WDYT ?

jbonofre replied:
After discussing with @davorbonaci I will move the version in the dependencyManagement of the main parent POM.

davorbonaci commented on sdks/java/io/jms/pom.xml:
```diff
@@ -0,0 +1,98 @@
+    <dependency>
+      <groupId>org.apache.activemq</groupId>
+      <artifactId>activemq-kahadb-store</artifactId>
+      <version>5.11.2</version>
```
Move version to the top-level?

jbonofre replied:
As it's JMS specific and for the test, I think it makes sense to keep it local to the IO. WDYT ?

It's the same as for the KafkaIO for instance.

davorbonaci commented on sdks/java/io/jms/pom.xml:
```diff
@@ -0,0 +1,98 @@
+    </plugins>
+  </build>
+
+  <dependencies>
```
We should analyze the new dependencies, for license-compliance, size, stability, etc.

Can you make sure we log this in the review by pasting the output of the **verbose** dependency tree and licenses.

jbonofre replied:
Good idea, let me prepare this. Maybe we can add a note in the github template about that (for new PR).

dhalperi replied:
JB: still todo?

jbonofre replied:
I checked the dependencies license and it's all Apache (as we are using Apache projects ;)).

davorbonaci commented:
I'll leave the rest of the review for @dhalperi.

jbonofre commented:
Rebased and updated to move the dependencies definition in the top level pom.

About the license, the JmsIO dependencies are Apache projects (Apache ActiveMQ and Apache Geronimo), so it's already under Apache license.

dhalperi commented on pom.xml:
```diff
@@ -573,6 +575,23 @@
+        <groupId>org.apache.activemq</groupId>
+        <artifactId>activemq-kahadb-store</artifactId>
+        <version>${activemq.version}</version>
+      </dependency>
```
I'm a little confused about this. I would think that these should just be dependencies of the JMS module rather than the entire project.

jbonofre replied:
I agree: it's what I proposed first. Davor said he would prefer on top level if some other parts use it.
However, as we did in Kafka, I would keep it "local" to the JMS IO module.
I can revert this.

dhalperi commented on sdks/java/io/jms/pom.xml:
```diff
@@ -0,0 +1,95 @@
+
+  <artifactId>jms</artifactId>
+  <name>Apache Beam :: SDKs :: Java :: IO :: JMS</name>
+  <description>IO to read and write to JMS destinations (queues and topics).</description>
```
Maybe expand JMS in one of these two locations?

jbonofre replied:
You mean be able to consume from multiple queues and topics (as we do in the Kafka IO) ?

dhalperi replied:
Just meant the "JMS" acronym itself. Not everyone obviously knows what JMS stands for :)

dhalperi commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsCheckpointMark.java:
```diff
@@ -0,0 +1,55 @@
+  }
+
+  @Override
+  public void finalizeCheckpoint() throws IOException {
```
drop the `throws`?

jbonofre replied:
+1 gonna do that.

dhalperi commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java:
```diff
@@ -0,0 +1,488 @@
+ * <p>JmsIO source returns unbounded collection of JMS records as {@code PCollection<JmsRecord<T>>}.
+ * A {@link JmsRecord} includes JMS headers and properties, along with the JMS message payload.</p>
+ *
+ * <p>To configure a JMS source, you have to provide a {@code javax.jms.ConnectionFactory}
```
optional: more `@link` instead of `@code` where  possible?

(recall that I'm being deliberately verbose in my comments to you ;)

jbonofre replied:
+1, thanks !

dhalperi commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java:
```diff
@@ -0,0 +1,488 @@
+        transform = unbounded.withMaxNumRecords(maxNumRecords);
+      } else if (maxReadTime != null) {
+        transform = unbounded.withMaxReadTime(maxReadTime);
+      }
```
should these always be mutually exclusive? What if the user wants both?

jbonofre replied:
Good point ;) I think we can deal with both defined in the same time.

dhalperi commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java:
```diff
@@ -0,0 +1,488 @@
+
+      PTransform<PInput, PCollection<JmsRecord>> transform = unbounded;
+
+      if (maxNumRecords < Long.MAX_VALUE) {
```
`!=`?

jbonofre replied:
+1

dhalperi commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java:
```diff
@@ -0,0 +1,488 @@
+      // handles unbounded source to bounded conversion if maxNumRecords is set.
+      Unbounded<JmsRecord> unbounded = org.apache.beam.sdk.io.Read.from(createSource());
+
+      PTransform<PInput, PCollection<JmsRecord>> transform = unbounded;
```
maybe just inline the declaration of unbounded here?

dhalperi commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java:
```diff
@@ -0,0 +1,488 @@
+      }
+      if (topic != null) {
+        builder.add(DisplayData.item("topic", topic));
+      }
```
R: @swegner to confirm, but I think this is roughly right.

Scott, is there some sort of `addIfNotNull` ?

dhalperi replied:
JB, do you want to add the class of the connection factory? Any connection information from the factory itself?

dhalperi replied:
maxNumRecords? MaxReadTime?

I think the function here is called `addIfNotDefault`

swegner replied:
> Scott, is there some sort of addIfNotNull ?

Yes: `builder.addIfNotNull(DisplayData.item("queue", queue));`

Note that you couldn't use this helper API in places where you dereference the possibly-null field in the call, i.e. `builder.addIfNotNull(DisplayData.item("foo", foo.toString()));`

dhalperi commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java:
```diff
@@ -0,0 +1,488 @@
+
+  private static class UnboundedJmsSource extends UnboundedSource<JmsRecord, JmsCheckpointMark> {
+
+    private final ConnectionFactory connectionFactory;
```
is the `connectionFactory` always guaranteed to be serializable?

jbonofre replied:
Finally, I double checked and according to the javadoc, JMS providers have to provide serializable `ConnectionFactory`.

dhalperi commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsCheckpointMark.java:
```diff
@@ -0,0 +1,55 @@
+ * Checkpoint for an unbounded JmsIO.Read. Consists of
+ * JMS destination name, and the latest message ID consumed so far.
+ */
+@DefaultCoder(SerializableCoder.class)
```
`SerializableCoder` is *strongly* discouraged in streaming pipelines / unbounded PCollections. It is very verbose (so that checkpoints and records are larger, increasing latency and the amount of data that has to be committed). It is also not guaranteed to be deserializable to an equal object in a different version of the JVM, so you might lose data if you restart a pipeline with an updated version of the code.

dhalperi replied:
In this case, I would recommend building a simple `StandardCoder` that uses `StringUtf8Coder` in a nested context to serialize the checkpoint mark.

dhalperi replied:
Actually, you could also use `@DefaultCoder(AvroCoder.class)` instead -- Avro can easily handle a pair of strings.

dhalperi commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsCheckpointMark.java:
```diff
@@ -0,0 +1,55 @@
+ * JMS destination name, and the latest message ID consumed so far.
+ */
+@DefaultCoder(SerializableCoder.class)
+public class JmsCheckpointMark implements UnboundedSource.CheckpointMark, Serializable {
```
(if you switch coders, you can remove the `Serializable` tag)

dhalperi commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java:
```diff
@@ -0,0 +1,488 @@
+  public static class Read extends PTransform<PBegin, PCollection<JmsRecord>> {
+
+    public Read withConnectionFactory(ConnectionFactory connectionFactory) {
+      return new Read(connectionFactory, queue, topic, maxNumRecords, maxReadTime);
```
totally optional: do you wish to do any validation here and below? e.g.,

* `checkNotNull(connectionFactory, "connectionFactory")`
* `checkArgument(maxNumRecords >= 0)`

dhalperi commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java:
```diff
@@ -0,0 +1,488 @@
+    @Override
+    public void validate(PBegin input) {
+      if (connectionFactory == null) {
+        throw new IllegalArgumentException("ConnectionFactory not specified");
```
could be simplified using a static import of `Preconditions.checkArgument`

jbonofre replied:
{{Preconditions}} from Google Base ?

dhalperi commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java:
```diff
@@ -0,0 +1,488 @@
+    private String queue;
+    private String topic;
+    private long maxNumRecords;
+    private Duration maxReadTime;
```
Can any of these be null? if so, please add `@Nullable` annotation here and in constructor.

dhalperi commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java:
```diff
@@ -0,0 +1,488 @@
+    @Override
+    public List<UnboundedJmsSource> generateInitialSplits(
+        int desiredNumSplits, PipelineOptions options) throws Exception {
+      return Collections.singletonList(new UnboundedJmsSource(connectionFactory, queue, topic));
```
`// TODO: implement splitting` ?

dhalperi commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java:
```diff
@@ -0,0 +1,488 @@
+      }
+      if (queue == null && topic == null) {
+        throw new IllegalArgumentException("Either queue or topic is not defined");
+      }
```
same (optional) comments re `checkArgument`

dhalperi commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java:
```diff
@@ -0,0 +1,488 @@
+
+    @Override
+    public Coder getCheckpointMarkCoder() {
+      return SerializableCoder.of(JmsCheckpointMark.class);
```
`AvroCoder.of(JmsCheckpointMark.class)` if you can switch the coder.

dhalperi commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java:
```diff
@@ -0,0 +1,488 @@
+
+    @Override
+    public Coder<JmsRecord> getDefaultOutputCoder() {
+      return SerializableCoder.of(JmsRecord.class);
```
I looked briefly at JmsRecord -- it seems hard to do much better than a Serializable coder or a JSON coder if you have a `Map<String, Object>` in the body.

dhalperi commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java:
```diff
@@ -0,0 +1,488 @@
+    private String destination;
+    private Connection connection;
+    private Session session;
+    private MessageConsumer consumer;
```
`@Nullable` on the member variables as appropriate?

dhalperi commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java:
```diff
@@ -0,0 +1,488 @@
+      try {
+        this.connection = connectionFactory.createConnection();
+        this.connection.start();
+        this.session = this.connection.createSession(false, Session.AUTO_ACKNOWLEDGE);
```
Can you comment on the `false`? `createSession(false /* no transacations */, Session.AUTO_ACKNOWLEDGE);`

But looking at the Connection javadoc:

> This method has been superseded by the method createSession(int sessionMode) which specifies the same information using a single argument, and by the method createSession() which is for use in a Java EE JTA transaction. Applications should consider using those methods instead of this one.

Maybe you just use this constructor instead? `https://docs.oracle.com/javaee/7/api/javax/jms/Connection.html#createSession-boolean-int-`

dhalperi replied:
Response to this suggestion?

jbonofre replied:
In the latest commit, I'm now using `CLIENT_ACK` and use the checkpoint mark to hold the messages and commit on finalize.
On the other hand, regarding `createSession` without the transaction argument, it depends of the JMS specification version. I would prefer to explicitly define the two arguments to be able to work with "all" JMS version.

dhalperi commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java:
```diff
@@ -0,0 +1,488 @@
+        this.connection.start();
+        this.session = this.connection.createSession(false, Session.AUTO_ACKNOWLEDGE);
+        if (source.topic != null) {
+          this.consumer = this.session.createConsumer(this.session.createTopic(source.topic));
```
again, very optional: you use a lot of `this.` when it's not strictly necessary. I tend to only use it in constructors where it is necessary:

```java
public myObject(Field field) {
    this.field = field;
}
```

The reason is that when it's not necessary I wonder if it's something something semantically meaningful, as in the case where an inner class wants to somehow call a function or variable from its parent (`this.parent.function()` eg)

dhalperi commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java:
```diff
@@ -0,0 +1,488 @@
+        headers.put("JMSRedelivered", message.getJMSRedelivered());
+        headers.put("JMSType", message.getJMSType());
+        headers.put("JMSExpiration", message.getJMSExpiration());
+        headers.put("JMSPriority", message.getJMSPriority());
```
Is it worth either:

1. extracting these well-known values into their own fields, instead of putting them in headers? [or having a JsmHeaders struct]?
2. Making these strings constants, such as `JMS_HEADER_MESSAGE_ID` so that it is easy for users to re-use them without fear of typos?

dhalperi commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java:
```diff
@@ -0,0 +1,488 @@
+        UnboundedJmsSource source,
+        @Nullable JmsCheckpointMark checkpointMark) {
+      this.source = source;
+      this.checkpointMark = checkpointMark;
```
`current = null`?

dhalperi commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java:
```diff
@@ -0,0 +1,488 @@
+        TextMessage message = (TextMessage) this.consumer.receiveNoWait();
+
+        if (message == null) {
+          return false;
```
before returning `false`, set `current` to `null`?

dhalperi replied:
ping above comment

This will make the source properly throw `NoSuchElementException` after advance has returned `false`.

jbonofre replied:
+1 thanks !

dhalperi commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java:
```diff
@@ -0,0 +1,488 @@
+
+    @Override
+    public JmsRecord getCurrent() throws NoSuchElementException {
+      return currentRecord;
```
```java
if (currentRecord == null) {
   throw new NoSuchElementException();
} else {
    return currentRecord;
}
```

dhalperi commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java:
```diff
@@ -0,0 +1,488 @@
+
+    @Override
+    public Instant getCurrentTimestamp() {
+      return currentTimestamp;
```
and similar here and above to getCurrent:

```java
if (currentRecord == null) {
   throw new NoSuchElementException();
} else {
    return currentTimestamp/Watermark;
}
```

dhalperi commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java:
```diff
@@ -0,0 +1,488 @@
+    @Override
+    public CheckpointMark getCheckpointMark() {
+      return new JmsCheckpointMark(destination,
+          currentRecord.getHeaders().get("JMSMessageID").toString());
```
another point in favor of either extracting these as fields (currentRecord.headers.messageId) or string constants :)

dhalperi commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java:
```diff
@@ -0,0 +1,488 @@
+      ConnectionFactory connectionFactory = source.connectionFactory;
+      try {
+        this.connection = connectionFactory.createConnection();
+        this.connection.start();
```
do you somehow need to pass in the message ID from the checkPointMark? what about the destination?

dhalperi commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java:
```diff
@@ -0,0 +1,488 @@
+          this.destination = source.topic;
+        } else {
+          this.consumer = this.session.createConsumer(this.session.createQueue(source.queue));
+          this.destination = source.queue;
```
It looks to me like `destination` comes always from the source, so you don't need it in the checkpoint mark.

The checkpoint mark is always used paired with the source a reader was using when it was created. So you only need record things that are not in the source.

jbonofre replied:
+1
Now the checkpoint stores a list of `Message`. It's the "pending" messages waiting to be ack.

dhalperi commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java:
```diff
@@ -0,0 +1,488 @@
+    public void close() throws IOException {
+      try {
+        if (consumer != null) {
+          consumer.close();
```
set it to `null` after so we don't close twice if this close is called twice for some reason? (here and below)

dhalperi commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java:
```diff
@@ -0,0 +1,488 @@
+        String value = ctx.element();
+
+        TextMessage message = session.createTextMessage(value);
+        producer.send(message);
```
currently in Beam we do not call finish bundle if there is an exception in this code.

To prevent connection leaks, maybe you want wrap this code as follows?

```java
try {
    TextMessage message = ...
    producer.send ...
} catch (RuntimeException | Exception t) {
    finishBundle(null);
    throw t;
}
```

or something like that.

@kennknowles he happens to not use the context in `finishBundle`, but is it easy to turn the `ProcessContext` into a `Context`?

dhalperi commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java:
```diff
@@ -0,0 +1,488 @@
+      public void finishBundle(Context c) throws Exception {
+        producer.close();
+        session.close();
+        connection.close();
```
set these to null after closing. That way if we reuse the DoFn we will re-create them.

dhalperi commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsRecord.java:
```diff
@@ -0,0 +1,72 @@
+
+  @Override
+  public int hashCode() {
+    return Arrays.deepHashCode(new Object[]{headers, properties, text});
```
Any reason to not simply use Java's [`Objects.hash(headers, properties, text)`](https://docs.oracle.com/javase/7/docs/api/java/util/Objects.html#hash(java.lang.Object...))?

dhalperi commented on sdks/java/io/pom.xml:
```diff
@@ -38,6 +38,7 @@
     <module>google-cloud-platform</module>
     <module>hdfs</module>
     <module>kafka</module>
+    <module>jms</module>
```
`j` before `k` ? :)

dhalperi commented:
Okay, I think I'm done with my current pass :) Thanks JB!

jbonofre commented:
Rebased and updated.

dhalperi commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsCheckpointMark.java:
```diff
@@ -0,0 +1,52 @@
+  private final String destination;
+  private final String messageId;
+
+  public JmsCheckpointMark(String destination, String messageId) {
```
I believe that for Avro (and hence AvroCoder) you often need a zero-args constructor. Have you actually tested that checkpoints can be encoded and decoded?

dhalperi replied:
(should add a test if not)

dhalperi commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsCheckpointMark.java:
```diff
@@ -0,0 +1,52 @@
+  private final String messageId;
+
+  public JmsCheckpointMark(String destination, String messageId) {
+    this.destination = destination;
```
maybe verify that these values are non-null when creating a checkpoint -- if they are, the coder will fail.

```java
this.destination = checkNotNull(destination, "destination");
this.messageId = checkNotNull(messageId, "messageId");
```

jbonofre commented:
Rebase (now I won't rebase again, I will add new commits on the PR).
I have the following questions:
- from my understanding watermark is when the data is actually processed in the pipeline, compare to the timestamp which is when the data has been actually created. It allows us to deal with the skew (the gap between data creation time and data processing time). Correct ? So, in the JMS case, the timestamp is basically the message timestamp and the watermark is when the message is consumed from the broker.
- the checkpoint is an object holding the last data consumed. It's a kind of unique identifier for the message. In the case of JMS, I'm using the Message ID which is unique and generated by the broker. Correct ?
- I implemented the initial split reader using the executor recommendation.

On the other hand, I have the following issues to fix:
- Following your recommendations, I'm now using `checkNotNull()` and `checkArgument()` methods. However, it seems that the arguments are always null. I have to figure out why (probably a stupid mistake ;)).
- The `testWrite()` test works fine.

dhalperi commented on sdks/java/io/jms/src/test/java/org/apache/beam/sdk/io/jms/JmsIOTest.java:
```diff
@@ -0,0 +1,139 @@
+  }
+
+  @Test
+  @Category(RunnableOnService.class)
```
I don't think this is really `RunnableOnService`. I think you mean `NeedsRunner`?

E.g., [this test](https://github.com/apache/incubator-beam/blob/bde2a856f1652f6ebab7aae2d5e78eeb4683a15a/sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/FlatMapElementsTest.java#L59).

dhalperi commented:
Build is now failing because of the new maven-dependency-plugin. It wants you to explicitly list all dependencies such as slf4j, guava, etc.

```
[INFO] --- maven-dependency-plugin:2.10:analyze-only (default) @ jms ---
[WARNING] Used undeclared dependencies found:
[WARNING]    org.slf4j:slf4j-api:jar:1.7.14:compile
[WARNING]    joda-time:joda-time:jar:2.4:compile
[WARNING]    com.google.guava:guava:jar:19.0:compile
[WARNING]    org.apache.activemq:activemq-client:jar:5.13.1:test
[WARNING]    com.google.code.findbugs:jsr305:jar:3.0.1:compile
```

jbonofre commented:
OK gonna fix that.

dhalperi commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java:
```diff
@@ -0,0 +1,516 @@
+          return false;
+        }
+
+        Map<String, Object> headers = new HashMap<>();
```
Hi JB, maybe I'm missing something obvious but I don't think that `headers` is used any more?

jbonofre replied:
Correct, I clean it up.

jbonofre replied:
By the way, as the build worked on my machine using `clean install -Papache-release` and `verify`, it would be great to include dependencies check in the release profile by default. Else, it means that an user may not be able to predict a Jenkins build issue.

dhalperi replied:
This was very recently added, so it's probably just not in your branch.

Maybe rebase, but don't squash?

jbonofre replied:
Indeed rebase makes the build failed on my machine as well. The dependencies are coming from the sdk-all dependency.

dhalperi commented:
> from my understanding watermark is when the data is actually processed in the pipeline, compare to the timestamp which is when the data has been actually created. It allows us to deal with the skew (the gap between data creation time and data processing time). Correct ? So, in the JMS case, the timestamp is basically the message timestamp and the watermark is when the message is consumed from the broker.

Watermark is a (heuristic) lower bound on all future timestamps (in event time). You are using the message time as the event time, and using the time of the most recent message as the watermark.

If I recall our whiteboard discussion correctly, JMS will deliver messages in timestamp order and in a consistent order even during failures -- so this makes perfect sense: timestamps will only grow, and the last message time is a good watermark.

This is not, however, the processing time. If the pipeline falls behind, JMS timestamps will be well before the processing time. So it is good that you are not using processing time as the watermark!

> the checkpoint is an object holding the last data consumed. It's a kind of unique identifier for the message. In the case of JMS, I'm using the Message ID which is unique and generated by the broker. Correct ?

Again, per our discussion this seems exactly right!

> I implemented the initial split reader using the executor recommendation.

Looks great!

dhalperi commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsCheckpointMark.java:
```diff
@@ -0,0 +1,48 @@
+  }
+
+  public String getMessageId() {
+    return messageId;
```
Is this value being used anywhere in the Reader? It seems to me that you need to actually use this information to seek to the right position when you begin reading, for example -- or else you will be re-reading the entire stream from scratch.

dhalperi replied:
The way to test this is something like the tests of [`CountingSource`](https://github.com/apache/incubator-beam/blob/master/sdks/java/core/src/test/java/org/apache/beam/sdk/io/CountingSourceTest.java#L297)

- Read a bunch of elements
- Take a checkpoint
- Read a bunch more elements
- Finalize the checkpoint you took before
- Restart and read until the end

The test:
* Ensure that the union of all elements you read before taking the checkpoint + after restarting is exactly the set of all elements you want to produce. Note that elements read after the checkpoint was taken but before restarting are NOT included in the output.

jbonofre replied:
Actually, not sure it makes sense from a JMS perspective: one the message has been read from the broker (an ACK is sent by the client), the message is deleted on the broker side. So, if you stop reading and resume the read later, you will read only the non-read messages.

jbonofre commented:
Thanks to @dhalperi I now have a better understanding of the requirements:
1. I should not use the AUTO_ACK mode on the JMS Session. Instead, I should use CLIENT_ACK and use the checkpoint mark to store the "pending" messages. Then, in the checkpoint finalize method, I will explicitly send the ACK to the JMS broker.
2. The watermark depends of the use cases and user needs. So, a good improvement would be to use a JMS property where the user can store his actually useful watermark. If this property is not defined in the JMS message, then we can fallback to the JMS timestamp (containing the timestamp when the message has been produced on the JMS broker).
3. Right now, the ConnectionFactory is an attribute of the PTransform (Read). As PTransform can be distributed on the executors, the ConnectionFactory has to be serializable. As the ConnectionFactory is an interface, it's hard to guarantee that any implementation will provide a serializable ConnectionFactory. A potential enhancement would be to use a kind of factory of connection factory, the factory itself would be serializable.

Huge thanks to @dhalperi for his help !

dhalperi commented:
Any time!

jbonofre commented:
Regarding 3, according to http://docs.oracle.com/javaee/1.4/api/javax/jms/ConnectionFactory.html, we can read:
```
It is expected that JMS providers will provide the tools an administrator needs to create and configure administered objects in a JNDI namespace. JMS provider implementations of administered objects should be both javax.jndi.Referenceable and java.io.Serializable so that they can be stored in all JNDI naming contexts. In addition, it is recommended that these implementations follow the JavaBeansTM design patterns.
```

So a `ConnectionFactory` implementation is `Serializable`.

jbonofre commented:
Gonna fix the dependencies duplicate.

dhalperi commented:
[update: Per discussion this morning, JB is still working on watermarks.]

jbonofre commented:
Jenkins failure is not related to the PR: it failed on archetype when retrieving the maven-archetype-plugin from Central.

dhalperi commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsCheckpointMark.java:
```diff
@@ -34,6 +34,7 @@
 public class JmsCheckpointMark implements UnboundedSource.CheckpointMark {
 
   private final List<Message> messages = new ArrayList<>();
+  private long oldestPendingTimestamp = System.currentTimeMillis();
```
* Suggest storing this as a Joda-time `Instant`.
* Suggest using [`BoundedWindow.TIMESTAMP_MIN_VALUE`](https://github.com/apache/incubator-beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/BoundedWindow.java#L38) as this is earlier than all timestamps. If you boot up and read old messages (before `now`), you don't want to mark the first few messages as LATE!

dhalperi commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsCheckpointMark.java:
```diff
@@ -56,6 +64,7 @@ public void finalizeCheckpoint() {
       }
     }
     messages.clear();
+    oldestPendingTimestamp = System.currentTimeMillis();
```
Rather than being `Instant.now()`, maybe it would make more sense to be the maximum timestamp among all of the elements in the batch? This assumes that event time is in increasing order, but it's again better than now because there might be old messages still.

dhalperi commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java:
```diff
@@ -367,7 +367,7 @@ public JmsRecord getCurrent() throws NoSuchElementException {
     @Override
     public Instant getWatermark() {
-      return currentTimestamp;
+      return new Instant(checkpointMark.getOldestPendingTimestamp());
```
Conversion unnecessary if you store the timestamp as an instant.

dhalperi commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsCheckpointMark.java:
```diff
@@ -59,12 +63,15 @@ public void finalizeCheckpoint() {
       try {
         message.acknowledge();
+        Instant currentMessageTimestamp = new Instant(message.getJMSTimestamp());
+        if (currentMessageTimestamp.isAfter(oldestPendingTimestamp)) {
```
Add a comment here, something like:

```
// Acknowledge all outstanding messages. Since we believe that messages
// will be delivered in timestamp order, and acknowledged messages will not be retried,
// the newest message in this batch is a good lower bound for future messages.
```

(Could also use oldest message for a more conservative bound in case some messages do get retried anyway, but that's up to you.)

jbonofre replied:
Good idea.

dhalperi commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsCheckpointMark.java:
```diff
@@ -34,7 +37,7 @@
 public class JmsCheckpointMark implements UnboundedSource.CheckpointMark {
```
Do take a quick pass over this class and try to restrict the visible APIs as much as possible.

In general, anything that's an internal implementation detail should be `private` or at least package-visible (if needed for testing). This will make our lives much easier once we start having a backwards-compatible guarantee.

jbonofre replied:
+1. Fixing that.

dhalperi commented:
Merged!

kennknowles commented on sdks/java/io/jms/pom.xml:
```diff
@@ -0,0 +1,136 @@
+    <relativePath>../pom.xml</relativePath>
+  </parent>
+
+  <artifactId>jms</artifactId>
```
`beam-sdks-java-io-jms`?
